### DISCUSSION TITLE
How to use "default" value for new properties ("migrations")?

### DISCUSSION THREAD

**Discussion by Prinzhorn** [dynamoose/dynamoose#1099] [Q&A] [answered]
I _think_ this is basically #905, but I didn't read the full discussion. I'm just wondering how others handle this. I was expecting `default` to work the same as it does in Mongoose. But the docs say

> You can set a default value for an attribute that **will be applied upon save**

For me the main use-case of `default` has always been to do "migrations". So if I add a new property I don't have to worry if it exists in the DB because Mongoose would use the default if it didn't so the application would just work. That doesn't work with dynamoose.

Note: the code examples create a new object because it made the examples easier (doesn't require an actual db connection). In reality I want this to work with documents I fetch from the database so I can transparently use new properties without worrying.

## Mongoose

```js
const mongoose = require('mongoose');

const db = mongoose.connection;
const schema = new mongoose.Schema({ aNumber: { type: Number, default: 4.815162342 } });
const M = db.model('M', schema);
const m = new M();
console.log(m.aNumber); // 4.815162342
```

## Dynamoose

```js
const dynamoose = require('dynamoose');

const schema = new dynamoose.Schema({ aNumber: { type: Number, default: 4.815162342 } });
const M = dynamoose.model('M', schema);
const m = new M();
console.log(m.aNumber); // undefined
```

Even this doesn't work (explicitly defining `get` to handle this)

```js
const dynamoose = require('dynamoose');

const schema = new dynamoose.Schema({ aNumber: { type: Number, default: 4.815162342, get: (value) => value || 1337 } });
const M = dynamoose.model('M', schema);
const m = new M();
console.log(m.aNumber); // undefined
```

**So how do others handle this without writing application code to check for every property that has ever been added and applying defaults manually?**

**Comment by fishcharlie** [comment, marked as answer]
@Prinzhorn This is a duplicate of #905. Please read that thread for more information. There are a few workaround solutions you can find in that thread.

**Reply by Prinzhorn** [reply]
Thanks, I've looked into #905 a bit closer, but couldn't see any discussion or workaround for the use-case I'm describing: applying defaults to the result of `get`, `query` etc. So when retrieving a document from DynamoDB that is missing the properties. But maybe I didn't look close enough? I assume Dynamoose is internally creating new instances of the model with the JSON from DynamoDB, so a solution like a `createWithDefaults` method doesn't work as the creation is not in my control.

**Reply by fishcharlie** [reply]
@Prinzhorn Ahh got it. Yeah, nothing like that currently. You'd have to take the code in that answer, and adopt it to your own uses.

So the `createWithDefaults` you should be able to modify that somehow to have it work with `get` and `query`.

But this functionality isn't built into Dynamoose currently.

**Reply by Prinzhorn** [reply]
Thanks again for your quick help! I didn't follow the full discussion in #905 in detail. I'm sure there are good reasons why it currently works the way it does. I think when in doubt everything should work as it does in Mongoose, whenever appropriate (e.g. unless MongoDB and DynamoDB diverge in an incompatible way). A lot of thought has been put into Mongoose for it to reach v5, but my issues with Dynamoose probably mostly come from my lack of knowledge (and interest) in DynamoDB :smile:
